espnet/nets/beam_search_transducer.py

```diff
@@ -33,601 +31,634 @@ class Hypothesis:
+                y, state, lm_tokens = self.decoder.score(max_hyp, cache, init_tensor)
 
-                    hyps.append(new_hyp)
+                ytu = torch.log_softmax(self.decoder.joint_network(hi, y[0]), dim=-1)
 
-            hyps_max = float(max(hyps, key=lambda x: x.score).score)
-            kept_most_prob = sorted(
-                [hyp for hyp in kept_hyps if hyp.score > hyps_max],
-                key=lambda x: x.score,
-            )
-            if len(kept_most_prob) >= beam:
-                kept_hyps = kept_most_prob
-                break
+                top_k = ytu[1:].topk(beam_k, dim=-1)
 
-    if normscore:
-        nbest_hyps = sorted(
-            kept_hyps, key=lambda x: x.score / len(x.yseq), reverse=True
-        )[:nbest]
-    else:
-        nbest_hyps = sorted(kept_hyps, key=lambda x: x.score, reverse=True)[:nbest]
+                ytu = (
+                    torch.cat((top_k[0], ytu[0:1])),
+                    torch.cat((top_k[1] + 1, blank_tensor)),
+                )
 
-    return [asdict(n) for n in nbest_hyps]
+                if self.lm:
+                    lm_state, lm_scores = self.lm.predict(max_hyp.lm_state, lm_tokens)
 
+                for logp, k in zip(*ytu):
+                    new_hyp = Hypothesis(
+                        score=(max_hyp.score + float(logp)),
+                        yseq=max_hyp.yseq[:],
+                        dec_state=max_hyp.dec_state,
+                        lm_state=max_hyp.lm_state,
+                    )
 
-def time_sync_decoding(decoder, h, recog_args, rnnlm=None):
-    """Time synchronous beam search implementation.
+                    if k == self.blank:
+                        kept_hyps.append(new_hyp)
+                    else:
+                        new_hyp.dec_state = state
 
-    Based on https://ieeexplore.ieee.org/document/9053040
+                        new_hyp.yseq.append(int(k))
 
-    Args:
-        decoder (class): decoder class
-        h (torch.Tensor): encoder hidden state sequences (Tmax, Henc)
-        recog_args (Namespace): argument Namespace containing options
-        rnnlm (torch.nn.Module): language module
+                        if self.lm:
+                            new_hyp.lm_state = lm_state
+                            new_hyp.score += self.lm_weight * lm_scores[0][k]
 
-    Returns:
-        nbest_hyps (list of dicts): n-best decoding results
+                        hyps.append(new_hyp)
 
-    """
-    beam = min(recog_args.beam_size, decoder.odim)
+                hyps_max = float(max(hyps, key=lambda x: x.score).score)
+                kept_most_prob = sorted(
+                    [hyp for hyp in kept_hyps if hyp.score > hyps_max],
+                    key=lambda x: x.score,
+                )
+                if len(kept_most_prob) >= beam:
+                    kept_hyps = kept_most_prob
+                    break
 
-    max_sym_exp = recog_args.max_sym_exp
-    nbest = recog_args.nbest
+        if self.score_norm:
```

Comment:
I wonder if all the algorithms support this option. How about moving this if/else block into a new method like `def sort_nbest(self, hyps)` and reuse it in other algorithms.

Reply:
Default beam search and NSC use score normalization but it's an option for the first one while it's enforced in the second!